File: algorithms/92. Reverse Linked List II/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        start = dummy = ListNode(0)
        cur = dummy.next = head
        for i in xrange(1, m):
            cur = cur.next
            start = start.next
        pre = cur
        cur = cur.next
        for i in xrange(m, n):
            next = cur.next
            cur.next = pre
            pre = cur
            cur = next
        start.next.next = cur
        start.next = pre
        return dummy.next

# 44 / 44 test cases passed.
# Status: Accepted
# Runtime: 35 ms
# beats 97.75 %

File: algorithms/92. Reverse Linked List II/solution.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode dummy(0);
        dummy.next = head;
        auto pre = & dummy;
        for ( int i = 1; i < m; ++i )
            pre = pre->next;
        ListNode * rhead = pre->next, * rtail;
        for ( int i = 0; i < n - m; ++i ) {
            rtail = rhead->next;
            rhead->next = rtail->next;
            rtail->next = pre->next;
            pre->next = rtail;
        }
        return dummy.next;
    }
};
